website/blog/2023-02-10.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,504 @@
+  they are not arrays.
+
+```js
+db.mixedfilter.find({v:{$lt:['c']}})
+[
+  { _id: 'array-empty', v: [] },              // empty array is the smallest array
+  { _id: 'array-mixed', v: [ 0, 'a', 'z' ] }, // compares each element of [ 0, 'a', 'z' ]
+  { _id: 'array-numbers', v: [ 10, 5 ] }      // compares each element of [ 10, 5 ]
+]
+```
+
+Let's look at filtering greater than `['c']`.
+It selects array BSON type and compares the array against `['c']`.
+```js
+db.mixedfilter.find({v: {$gt: ['c']}})
+[ { _id: 'array-strings', v: [ 'd' ] } ] // compares each element of [ 'd' ]
+```
+#### Filtering nested array 
+Let's look at filtering nested array with an array.
+```js
+db.nestedfilter.insertMany([
+  { _id: 'nested-empty', v: [ [] ] },
+  { _id: 'nested-mixed', v: [ [ 'a', 1 ] ] },
+  { _id: 'nested-string', v: [ [ 'g', 'a' ] ] },
+  { _id: 'nested-number', v: [ [ 10, 1 ] ] },
+  { _id: 'array-bool', v: [ true ] },
+  { _id: 'array-null', v: [ null ] }
+])
+```
+Using non-array filter such as `'c'`, nothing was matched since `$gt` and `$lt`
+does not find string BSON type within the array, they are in the nested array.
+```js
+db.nestedfilter.find({v:{$lt: 'c'}})
+
+db.nestedfilter.find({v:{$gt: 'c'}})
+```
+
+##### $lt
+Using array filter `['c']` on `$lt` operator, it selected 
+empty nested array `{ _id: 'nested-empty', v: [ [] ] }`
+and `{ _id: 'nested-mixed', v: [ [ 'a', 1 ] ] }` which the nested array contains
+smaller element `a` than filter `c` on its first index.
+* selected `{ _id: 'nested-empty', v: [ [] ] }` because empty nested array is less.
+* selected `{ _id: 'nested-mixed', v: [ [ 'a', 1 ] ] }` because first index of 
+  nested array `a` is less than `c`.
+* selected ` { _id: 'nested-number', v: [ [ 10, 1 ] ] }` because first index of nested array
+ `10` is less than `c` in BSON comparison order.
+* selected `{ _id: 'array-null', v: [ null ] }` because `null` less than string `c`
+  in BSON comparison order.
+* not selecting `{ _id: 'nested-string', v: [ [ 'g', 'a' ] ] }` because `g` is not
+  less than `c`.
+* not selecting `{ _id: 'array-bool', v: [ true ] }` because boolean is not less than
+  string in BSON comparison order.
+```js
+db.nestedfilter.find({v:{$lt: ['c']}})
+[
+  { _id: 'nested-empty', v: [ [] ] },
+  { _id: 'nested-mixed', v: [ [ 'a', 1 ] ] },
+  { _id: 'nested-number', v: [ [ 10, 1 ] ] },
+  { _id: 'array-null', v: [ null ] }
+]
+```
+
+##### $gt
+Using array filter `['c']` on `$gt` operator, it selected all nested arrays
+from the collection. This is unexpected since empty array
+`{ _id: 'nested-empty', v: [ [] ] }` is expected to be smaller.
+:::this example shows uncertain result of $gt operator:::
+```js
+db.nestedfilter.find({v:{$gt: ['c']}})
+[
+  { _id: 'nested-empty', v: [ [] ] },
+  { _id: 'nested-mixed', v: [ [ 'a', 1 ] ] },
+  { _id: 'nested-string', v: [ [ 'g', 'a' ] ] },
+  { _id: 'nested-number', v: [ [ 10, 1 ] ] },
+  { _id: 'array-bool', v: [ true ] }
+]
```

Comment:
@b1ron do you understand what is going on here? It just I can't quite interpret the result. :see_no_evil:

Reply:
I think I have an explanation in the blog :hugs:

Reply:
@chilagrow Yes I see. So, when there are multiple elements of different types it doesn't necessarily compare them it just uses the ordering to determine if it matches. My only contention is that it does seem to match certain types and that's also confirmed [here](https://jira.mongodb.org/browse/SERVER-11444?focusedCommentId=492157&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-492157).

Reply:
Thanks, are you good with the explanation? It seems that the response might differ depending on version..